### ISSUE TITLE
All-to-all STDP synapse question

### ISSUE THREAD

**Issue by Crypto1993** [genn-team/genn#69]
Good evening,
I'm implementing some code for simulating the antennal lobes of a Drosophila melanogaster, adopting a scheme in which there are 500 Izhikevich neuron. Now, the modeling stage of neuron population was relatively easy (I've set several proper parameters in order to fulfill some requirements), but I have a doubt; since my aim is defining all-to-all STDP connections, I've written the following code:

`model.addSynapsePopulation("Pop1self",LEARN1SYNAPSE,ALLTOALL,INDIVIDUALG,NO_DELAY,EXPDECAY,"Pop1","Pop1",_InitSynapse,_SynapseParameters,_InitPostSynaptic,_PostSynapticVariables);`

but my question is: are the keywords LEARN1SYNAPSE and ALLTOALL sufficient to accomplish this task? In particular, do I have to define something else for implementing the connectivity scheme I desire (for example, inside the .cu simulation file)?

Thanks in advance for your answers.

**Comment by tnowotny** [comment]
Hi Crypto1993,
whenever you specify INDIVIDUALG (which is the right thing to do here as you would want synapses to be able to have all different conductances, which also can change over time), you will have to provide  matrices of initial values for the internal variables of the synapses, i.e. for LEARN1SYNAPSE, this is mainly the "raw conductance" gRaw, and the resulting conductance "g". Have a look at the MBody1 example, function "read_kcdnsyns(...)" in file map_classol.cc for an example how this can look like.
Hope this helps.

**Comment by Crypto1993** [comment]
Good morning,
I had a look at the file you cited, but why do I have to set "again" initial conditions for the synapses state variables if I've already created a vector _InitSynapse which I use when I call the addSynapsePopulation function?

And moreover: what does it change if I simply use a connectivity scheme like the TenHH tutorial one mentioned in the reference manual?

`	// define the connectivity
	int pre, post;
	for (int i = 0; i < 500; i++) {
		pre = i;
		post = (i + 1) % 10;
		cout << "Synapses: "<< pre * 500 + post << endl;
		gPop1self[pre * 500 + post] = 0.01;
	}
	pushPop1selfStateToDevice();`

Surely for an all-to-all connectivity the previous code is wrong, but now I'd like to ask you if this procedure (i.e. creating the matrix inside the .cu file) is right.

Thanks for your help... and sorry for the insistence!

**Comment by jamesturner246** [comment]
Hi all.

I think this is worth considering further. Users should at least have the option of declaring connectivity matrices up-front, rather than always having to do it in user code. One could generate the matrix in the modeDefinition() function, via the *_ini variable, and use the values when code generating the initialize() function. I don't think it would be too much of a pain to change, either.

James,

**Comment by Crypto1993** [comment]
Since my aim is an all-to-all connectivity, I started using a simplified neuron population (just 2 neurons); because I use the option INDIVIDUALG, GeNN should create an array of 2x2=4 elements in which I can define my conductances (if I've understood well). I've assigned my conductances (I suppose it constant for each synapse) to each element of the vector, by using the instruction:
`gPop1self[i] = XXXX`
where XXXX is the desired value and i varies between 0 and 3. Instead of placing the previous instruction inside the modelDefinition function, I've written it inside the .cu file, as explained in several examples in the reference manual.

It seems too simple and strange, anyway, also because if I plot the membrane potentials of both neurons they are practically equal.

**Comment by jamesturner246** [comment]
Did you remember to copy the g values over with pushPop1selfStateToDevice()? Could I please take a look at your usercode?

**Comment by Crypto1993** [comment]
This is the .cc file:

`#define DT	0.01

#include "C:\genn-team-genn-8cacb78\lib\include\modelSpec.cc"
#include "C:\genn-team-genn-8cacb78\lib\include\modelSpec.h"
			
void modelDefinition(NNmodel &model){
		initGeNN();
		model.setName("Antennal_Lobes");
		
		double _InitNeuron[2]{
			-70.0,
			-14.0
		};
		
		double _NeuronParameters[4]{
			0.0,
			-0.1,
			-55.0,
			0.0
		};
			
		double _InitSynapse[2]{
			0.02,
			0.01
		};
			
		double _SynapseParameters[11]{
			50.0,
			50.0,
			50000.0,
			100000.0,
			200.0,
			0.015,
			0.0075,
			33.33,
			10.0,
			0.00006
		};
		
		double * _InitPostSynaptic = NULL;
		
		double _PostSynapticVariables[2]{
			1.0,
			0.0
		};
		
		model.addNeuronPopulation("Pop1",2,IZHIKEVICH,_NeuronParameters,_InitNeuron);
		model.addSynapsePopulation("Pop1self",LEARN1SYNAPSE,ALLTOALL,INDIVIDUALG,NO_DELAY,EXPDECAY,"Pop1","Pop1",_InitSynapse,_SynapseParameters,_InitPostSynaptic,_PostSynapticVariables);
		model.activateDirectInput("Pop1",CONSTINP);
		model.setConstInp("Pop1",10.0);
		model.finalize();
		
}`

while this is the .cu file:

`#include "AL.cc"
#include "Antennal_Lobes_CODE/runner.cc"
#include "C:\genn-team-genn-8cacb78\lib\include\hr_time.cpp"

int main()
{
	CStopWatch timer;
	timer.startTimer();
	allocateMem();
	initialize();
	
	for(int i = 0; i < 4; i++) gPop1self[i] = 0.02;
	pushPop1selfToDevice();
	
	ofstream os("output.V.dat");
	
	system("cls");
	cout << "ANTENNAL LOBE SIMULATION" << endl;
	cout << "- No. of neurons: 2" << endl;
	cout << "- Size of synapses array: 4" << endl;
	
	for (int i = 0; i < 4; i++) cout << "Synapse no. " << i << ": " << gPop1self[i] << endl;
		
	for (double t = 0; t < 10; t+=0.01) {
		stepTimeGPU(0.1);
		pullPop1StateFromDevice();
		os << t << " ";
		for (int j= 0; j < 2; j++) 	os << VPop1[j] << " ";
		os << endl;
	}
	
	cout << endl;
	os.close();
	timer.stopTimer();
	
	cout << "The simulation time is: " << timer.getElapsedTime() << " seconds " << endl;
	cout << "----------------------" << endl;
	return 0;
}`

**Comment by jamesturner246** [comment]
OK, I've had a look, and this seems to be correct GeNN behaviour. In your model, you initialise both neurons to -70 mV, and the conductance of all four synapses to 0.02 g - i.e. both neurons are getting the exact same spikes with the exact same strength. Even if you vary the conductances a little bit, both neurons are still getting roughly the same incoming current due to self-to-self connections.

Try varying the conductances a bit more; something like:

  gPop1self[0] = 0.09;
  gPop1self[1] = 0.15;
  gPop1self[2] = 0.03;
  gPop1self[3] = 0.02;

You may also get better divergence by increasing the numerical precision GeNN uses.

All the best,
James.

**Comment by Crypto1993** [comment]
Ok, @jamesturner246 , thanks for your clarification!
But there's still only one thing I don't understand: if I've specified a vector in which I insert initial conditions for the state variables, why do I have to re-insert them like you have done in your previous post?

**Comment by jamesturner246** [comment]
If you want all synapses in the group to have the same g value, then you don't have to do it again. The g parameter in your modelDefinition() function is assigned to all synapses in the group, and you don't have to modify and copy them over in user code. 

But if you want different synapses within the same group to be initialised with different g values, then *that* is when you have to redefine them in user code, and then copy them over. I will discuss with my colleges the option of defining a synapse group with mixed g values all inside the modelDefinition() function, so noone will have to do it in their own usercode if they don't want to!

All the best,
James.

**Comment by jamesturner246** [comment]
I think I might know what is confusing you. The 0 index of your _InitSynapse array is the global *g* value, to be used as the g value for all synapses in the group. The 1 index of _InitSynapse array is the global *graw* parameter (the raw conductance - ask Thomas about this). It is not the conductance *g* of the second neuron in the group!

**Comment by Crypto1993** [comment]
Yes, I've understood it studying the reference manual. I've look at the syntax and the order of parameters required for implementing STDP synapses.

I would you thank you again for your help... and sorry for my late reply!

**Comment by tnowotny** [comment]
Hi - I should add that the provided STDP synapses are a bit unusual in the following way: There are two internal variables that reflect the synaptic conductance: g and graw. They are directly related to each other (basically g is graw filtered through a sigmoid function; g= sigmoid(graw)), so one of the two variables could be omitted; but having both saves a lot of computation for not having to calculate the sigmoid all the time.
If g and graw are not initialized to matching values, e.g. if one is initialized and the other isn't, then it's eventually graw that will prevail (at each spike event, g is re-initialized from graw).
In summary, for these STDP synapses, both g and graw need to be set to values that correspond to each other (g = sigmoid(graw)) in order for this to work well.

Note to developers: a) we may want to consider withdrawing this predefined synapse type as it's too confusing. Probably should add some more standard model with a hard limit on g rather than the g/graw filter solution.
b) even for this synapse type, if it is retained, we may want to consider to discontinue g as a variable and take the performance hit of computing the sigmoid filter repeatedly - has anyone benchmarked this?

**Comment by Crypto1993** [comment]
Ok, got it.
Another doubt: I know vector of synapses between a population of n pre-synaptic neurons and m post-synaptic neurons is made of nxm elements... but if I pick the i-th element of gPop1self vector, what does it represent? Which neurons does it link?

**Comment by tnowotny** [comment]
It's organized pre-synaptic index major, post-synaptic index minor, i.e.
the variable for a synapse from neuron i to neuron j is located at
k= i * n_post + j
Inverting is best done with integer division and modulo:
i= k div n_post
j= k mod n_post

**Comment by Crypto1993** [comment]
Thank you!

**Comment by jamesturner246** [comment]
That's correct.

**Comment by Crypto1993** [comment]
Stupid question: let us suppose I want to define ALL-TO-ALL connectivity among the neurons of a population of N elements. Basing on what I read, vector of g conductances will be of N^2 elements... but mathematically when one element is related to all the others, the number of links is only N*(N-1)/2. Now, the question is: what do the other N(N+1)/2 (equal to N^2 - N(N-1)/2) elements represent?

**Comment by tnowotny** [comment]
N(N-1)/2 is the number of pairs. You could interpret it as the number of **undirected** connections. If you have both directions, it's 2*N(N-1)/2= N(N-1) and if you allow neurons to connect to themselves it's N(N-1)+N= N^2.
